runtime/j9jitbuilder/JavaMethodCallbackGen.cpp

```diff
@@ -19,23 +19,303 @@
+	if (NULL == _paramIdentifiers) {
+		rc == OOM;
+	} else {
+		DefineLine(LINETOSTR(__LINE__));
+		DefineFile(__FILE__);
+
+		/* Define all parameter type names in _paramIdentifiers one after each other. Since we don't know
+		 * the names we will use numbers instead. The max number of method parameters that Java allows is
+		 * 255, so parameters will be named 000\0 001\0, 002\0 ... 255\0. Each parameter name will be
+		 * 4bytes including NULL terminator.
+		 */
+		for (UDATA i = 0; i < _paramTypeLen; i++) {
+			snprintf((char*)(_paramIdentifiers + i), sizeof(U_32),"%.3u\0", i);
+			DefineParameter((char*)(_paramIdentifiers + i), getILTypeFromJ9Class(_paramTypes[i]));
+		}
+
+		/* define the JNI call-in function. Helpers look at the
+		 * _returnType to determine what kind of JNI call-in function
+		 * to use. Call-in function will use the
+		 * Call[type]MethodA(JNIEnv *env, jobject obj, jmethodID methodID, const jvalue *args)
+		 * JNI function so it will only take 4 parameters.
+		 */
+		DefineName("JavaMethodCallbackGen");
+		DefineReturnType(getILTypeFromJ9Class(returnType));
+
+		DefineFunction((char *)getCallInFunctionName(),
+		                  (char *)"jni.h",
+		                  getCallInFunctionLineNumber(),
+		                  getCallInFunctionAddress(),
+						  getILTypeFromJ9Class(_returnType),
+						  4,
+						  Address,
+						  Address,
+						  Address,
+						  Address);
+
+		/* Define the JNI jvalue union type */
+		if (_paramTypeLen > 0) {
+			_dictionary->DefineUnion("jvalue");
+			_dictionary->UnionField("jvalue", "Z", _dictionary->toIlType<uint8_t>());
+			_dictionary->UnionField("jvalue", "B", Int8);
+			_dictionary->UnionField("jvalue", "C", _dictionary->toIlType<uint16_t>());
+			_dictionary->UnionField("jvalue", "S", Int16);
+			_dictionary->UnionField("jvalue", "I", Int32);
+			_dictionary->UnionField("jvalue", "J", Int64);
+			_dictionary->UnionField("jvalue", "F", Float);
+			_dictionary->UnionField("jvalue", "D", Double);
+			_dictionary->UnionField("jvalue", "L", Address);
+			_dictionary->CloseUnion("jvalue");
+		}
+	}
+
+	return rc;
+}
+
+void
+JavaMethodCallbackGen::destroy()
+{
+	PORT_ACCESS_FROM_VMC(_vmThread);
+	j9mem_free_memory(_paramIdentifiers);
 }
 
 bool
 JavaMethodCallbackGen::buildIL()
 {
+	if (_paramTypeLen > 0) {
+		TR::IlType *jValueType = _dictionary->LookupUnion("jvalue");
+		TR::IlType *pJvalue = _dictionary->PointerTo(jValueType);
+
+		/* create array of jvalues (length of parameters) on stack */
+		Store("parameters",
+				CreateLocalArray(_paramTypeLen, jValueType));
+
+		/* take each function parameter and load it into the jvalue array (parameters) */
+		for (UDATA i = 0; i < _paramTypeLen; i++) {
+			/* get pointer to jvalue at parameter index */
+			Store("param", IndexAt(pJvalue,
+			         Load("parameters"),
+			         ConstInt32(i)));
+
+			/* Fill jvalue with zero first. jvalue is 64bits */
+			StoreIndirect("jvalue", "J",
+					Load("param"),
+					ConstInt64(0));
```

Comment:
Is ConstInt64 the correct type?  Does this require a ConvertTo call?

Reply:
The field type for for `"J"` is a `Int64`, so I would assume that the types match up and no conversion is required.

I am trying to do the equivalent of this:
```
jvalue *param = ...;
param->J = 0L; //where J is an int64_t
```

Reply:
Sounds good